pkg/distributor/forwarding/forwarding_test.go

```diff
@@ -770,6 +775,161 @@ func BenchmarkRemoteWriteForwarding(b *testing.B) {
+
+	for _, req := range append(reqs1(), reqs2()...) {
+		ce := ""
+		ct := ""
+
+		for _, h := range req.Headers {
+			if h.Key == "Content-Encoding" {
+				ce = h.Values[0]
+			}
+			if h.Key == "Content-Type" {
+				ct = h.Values[0]
+			}
+		}
+		require.Equal(t, "snappy", ce)
+		require.Equal(t, "application/x-protobuf", ct)
+	}
+
+	bodies := reqs1()
+	require.Len(t, bodies, 1)
+	receivedReq1 := decodeBody(t, bodies[0].Body)
+	require.Len(t, receivedReq1.Timeseries, 2)
+	requireLabelsEqual(t, receivedReq1.Timeseries[0].Labels, "__name__", "metric1", "some_label", "foo")
+	requireSamplesEqual(t, receivedReq1.Timeseries[0].Samples, now, 1)
+	require.Empty(t, receivedReq1.Timeseries[0].Exemplars)
+	requireLabelsEqual(t, receivedReq1.Timeseries[1].Labels, "__name__", "metric1", "some_label", "bar")
+	requireSamplesEqual(t, receivedReq1.Timeseries[1].Samples, now, 2)
+	require.Empty(t, receivedReq1.Timeseries[1].Exemplars)
+
+	bodies = reqs2()
+	require.Len(t, bodies, 1)
+	receivedReq2 := decodeBody(t, bodies[0].Body)
+	require.Len(t, receivedReq2.Timeseries, 2)
+	requireLabelsEqual(t, receivedReq2.Timeseries[0].Labels, "__name__", "metric2", "some_label", "foo")
+	requireSamplesEqual(t, receivedReq2.Timeseries[0].Samples, now, 3)
+	require.Empty(t, receivedReq2.Timeseries[0].Exemplars)
+	requireLabelsEqual(t, receivedReq2.Timeseries[1].Labels, "__name__", "metric2", "some_label", "bar")
+	requireSamplesEqual(t, receivedReq2.Timeseries[1].Samples, now, 4)
+	require.Empty(t, receivedReq2.Timeseries[1].Exemplars)
+
+	expectedMetrics := `
+	# HELP cortex_distributor_forward_requests_total The total number of requests the Distributor made to forward samples.
+	# TYPE cortex_distributor_forward_requests_total counter
+	cortex_distributor_forward_requests_total{} 2
+	# HELP cortex_distributor_forward_samples_total The total number of samples the Distributor forwarded.
+	# TYPE cortex_distributor_forward_samples_total counter
+	cortex_distributor_forward_samples_total{} 4
+	# TYPE cortex_distributor_forward_grpc_clients gauge
+	# HELP cortex_distributor_forward_grpc_clients Number of gRPC clients used by Distributor forwarder.
+	cortex_distributor_forward_grpc_clients 2
+`
+
+	require.NoError(t, testutil.GatherAndCompare(
+		reg,
+		strings.NewReader(expectedMetrics),
+		"cortex_distributor_forward_requests_total",
+		"cortex_distributor_forward_samples_total",
+		"cortex_distributor_forward_grpc_clients",
+	))
+}
+
+func newTestHttpgrpcServer(tb testing.TB, status int) (string, func() []*httpgrpc.HTTPRequest) {
+	tb.Helper()
+
+	server := grpc.NewServer()
+
+	// Register handler for httpgrpc
+	hs := &httpgrpcServer{status: int32(status)}
+	httpgrpc.RegisterHTTPServer(server, hs)
+
+	// Register health server (pool requires health checks to pass).
+	grpc_health_v1.RegisterHealthServer(server, healthServer{})
+
+	l, err := net.Listen("tcp", "")
+	require.NoError(tb, err)
+
+	go func() {
+		_ = server.Serve(l)
+	}()
+	tb.Cleanup(func() {
+		_ = l.Close()
```

Comment:
shouldn't the server also be stopped?

```
		server.Stop()
```

Reply:
grpc server stops when listener is closed, but calling `server.Stop()` explicitly doesn't hurt. Adding it.